### PR TITLE
Number_types:  Add Exact_algebraic

### LINKED ISSUE #3702: Remaining issues of the testsuite with LEDA instead of GMP
## Issue Details

We want to be able to replace `GMP` and `CGAL_Core` with `leda-numbers`. However there are many `CMakeLists.txt` testing for `CGAL_Core` explicitly...

### `CMakeLists.txt` reffering to CGAL_Core
```
Arrangement_on_surface_2/demo/Arrangement_on_surface_2/CMakeLists.txt
CGAL_Core/examples/Core/CMakeLists.txt
CGAL_Core/src/CGAL_Core/CMakeLists.txt
Cone_spanners_2/examples/Cone_spanners_2/CMakeLists.txt
Cone_spanners_2/test/Cone_spanners_2/CMakeLists.txt
GraphicsView/demo/Polygon/CMakeLists.txt
Kernel_23/test/Kernel_23/CMakeLists.txt
Surface_mesh_shortest_path/test/Surface_mesh_shortest_path/CMakeLists.txt
Surface_sweep_2/test/Surface_sweep_2/CMakeLists.txt
```

It should be fixed, so that the tests are actually compiled.

### Abstraction of Exact_nt_with_sqrt

Maybe `CGALConfig.cmake` could provide a CMake target `CGAL::Exact_nt_with_sqrt` that would be an abstraction of `CGAL_Core` and LEDA.

### Remaining issues in the testsuite
See https://github.com/CGAL/cgal/pull/3501#issuecomment-444522806
> warnings and errors remaing on platforms with LEDA:
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Algebraic_kernel_d
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Kernel_23
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Polygon_mesh_processing
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Arrangement_on_surface_2
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Bounding_volumes
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Envelope_3
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#NewKernel_d
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#QP_solver
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Segment_Delaunay_graph_Linf_2
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Segment_Delaunay_graph_Linf_2_Examples
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Triangulation
- [ ] https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Visibility_2
## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): all 
* Compiler: all 
* Release or debug mode: all 
* CGAL version: `master`/4.14

See the original discussion in https://github.com/CGAL/cgal/pull/3501.

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

Introduce an abstraction of `CORE::Expr` and `leda_real` to simplify tests and examples and to also test with Leda as now Core is often hardwired.   @lrineau will fix the CMakeLists.txt files.

## Release Management

* Affected package(s): see the issue 
* Issue(s) solved (if any): fix #3702

### CODE REVIEW

**Comment by sloriot** [comment]
`Exact_real` looks better and is more verbose IMO.

**Comment by afabri** [comment]
> 
> 
> `Exact_real` looks better and is more verbose IMO.

See #3702

File: Installation/cmake/modules/CGAL_SetupCGAL_CoreDependencies.cmake [hunk 1]
```diff
@@ -59,6 +59,14 @@ if (NOT MSVC)
   find_package( Boost 1.48 REQUIRED COMPONENTS thread system )
 endif()
 
+function(CGAL_Core_create_alias_Exact_rational_support)
+  if(NOT TARGET CGAL::Exact_rational_support)
```
Comment (lrineau) [inline Installation/cmake/modules/CGAL_SetupCGAL_CoreDependencies.cmake]: I have reverted the commit.

But anyway, the name should have been `Exact_algebraic_support`.

File: Hyperbolic_triangulation_2/demo/Hyperbolic_triangulation_2/CMakeLists.txt [hunk 2]
```diff
@@ -9,12 +9,12 @@ set(CMAKE_AUTOMOC ON)
...
 
 find_package(CGAL QUIET COMPONENTS Core Qt5)
-
+find_package(LEDA QUIET)
 find_package(Qt5 QUIET COMPONENTS Widgets)
 
 include_directories (BEFORE include)
 
-if(CGAL_FOUND AND CGAL_Qt5_FOUND AND Qt5_FOUND AND CGAL_Core_FOUND)
+if(CGAL_FOUND AND (CGAL_Qt5_FOUND OR LEDA_FOUND) AND Qt5_FOUND AND CGAL_Core_FOUND)
```
Comment (lrineau) [inline Hyperbolic_triangulation_2/demo/Hyperbolic_triangulation_2/CMakeLists.txt]: The `OR LEDA_FOUND` is not placed at the right place.

Comment (lrineau) [reply Hyperbolic_triangulation_2/demo/Hyperbolic_triangulation_2/CMakeLists.txt]: Well, you have already fixed in a later commit.

**Comment by sloriot** [comment]
@afabri is this PR ready for testing?

**Comment by sloriot** [comment]
+ conflicts

**Comment by afabri** [comment]
@maxGimeno conflicts are resolved.

File: Segment_Delaunay_graph_2/test/Segment_Delaunay_graph_2/test_sdg_2_et.cpp [hunk 3]
```diff
@@ -1,54 +1,17 @@
-#include <CGAL/basic.h>
```
Comment (lrineau) [inline Segment_Delaunay_graph_2/test/Segment_Delaunay_graph_2/test_sdg_2_et.cpp]: This commit is responsible for the compilation errors in:
https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-26.shtml#Segment_Delaunay_graph_2

Comment (afabri) [reply Segment_Delaunay_graph_2/test/Segment_Delaunay_graph_2/test_sdg_2_et.cpp]: We should test it again. I want to check that the error only happens in the draw functions, and if so fix that after a successfull testsuite without calling draw functions.

**Comment by lrineau** [comment]
@afabri: you have [forced-pushed](https://github.com/CGAL/cgal/compare/1807d2490ec77d05ca3f69d4bbb348f695186311..c738d89acaf68f0dcd732f269d296744d73bee84), and thus you have erased by commit https://github.com/CGAL/cgal/commit/1807d2490ec77d05ca3f69d4bbb348f695186311. I will rebase on your branch and repush the commit, and I hope you will not do that again.

File: Segment_Delaunay_graph_2/test/Segment_Delaunay_graph_2/test_sdg_2_et.cpp (hunk 3 above)
Comment (lrineau) [reply Segment_Delaunay_graph_2/test/Segment_Delaunay_graph_2/test_sdg_2_et.cpp]: What gave the tests on your machine?

Comment (lrineau) [reply Segment_Delaunay_graph_2/test/Segment_Delaunay_graph_2/test_sdg_2_et.cpp]: On my Linux machine, after commits https://github.com/CGAL/cgal/pull/3797/commits/c738d89acaf68f0dcd732f269d296744d73bee84 and https://github.com/CGAL/cgal/pull/3797/commits/7ffbf131bb615e19699680daa2e739ffb123d066, the following tests passed:
```shell-session
$ make && ctest -R 'test_sdg_2_et|test_sdg_hierarchy_2_et' -j2
    Start 3812: compilation_of__test_sdg_2_et
    Start 3814: Segment_Delaunay_graph_2_Tests_SetupFixture
1/6 Test #3814: Segment_Delaunay_graph_2_Tests_SetupFixture .....   Passed    0.01 sec
    Start 3818: compilation_of__test_sdg_hierarchy_2_et
2/6 Test #3818: compilation_of__test_sdg_hierarchy_2_et .........   Passed   15.22 sec
    Start 3819: execution___of__test_sdg_hierarchy_2_et
3/6 Test #3812: compilation_of__test_sdg_2_et ...................   Passed   15.82 sec
    Start 3813: execution___of__test_sdg_2_et
4/6 Test #3819: execution___of__test_sdg_hierarchy_2_et .........   Passed    0.83 sec
5/6 Test #3813: execution___of__test_sdg_2_et ...................   Passed    1.11 sec
    Start 3815: Segment_Delaunay_graph_2_Tests_CleanupFixture
6/6 Test #3815: Segment_Delaunay_graph_2_Tests_CleanupFixture ...   Passed    0.01 sec

100% tests passed, 0 tests failed out of 6

Label Time Summary:
Segment_Delaunay_graph_2_Tests    =  33.00 sec*proc (6 tests)

Total Test time (real) =  17.10 sec
30.40s user 2.40s system 191% cpu 17.109s total
```

**Comment by afabri** [comment]
@maxGimeno please add it to integration.

**Comment by afabri** [comment]
> 
> 
> @afabri: you have [forced-pushed](https://github.com/CGAL/cgal/pull/3797#event-2314230773), and thus you have erased by commit [1807d24](https://github.com/CGAL/cgal/commit/1807d2490ec77d05ca3f69d4bbb348f695186311). I will rebase on your branch and repush the commit, and I hope you will not do that again.

@lrineau Sorry, but I was sure that I had pulled your change.  Did you re-push?

**Comment by lrineau** [comment]
[**@afabri**](https://github.com/afabri) commented on [May 6, 2019, 9:29 AM GMT+2](https://github.com/CGAL/cgal/pull/3797#issuecomment-489524612 "2019-05-06T07:29:55Z - Replied by Github Reply Comments"):
> > 
> > 
> > @afabri: you have [forced-pushed](https://github.com/CGAL/cgal/pull/3797#event-2314230773), and thus you have erased by commit [1807d24](https://github.com/CGAL/cgal/commit/1807d2490ec77d05ca3f69d4bbb348f695186311). I will rebase on your branch and repush the commit, and I hope you will not do that again.
> 
> @lrineau Sorry, but I was sure that I had pulled your change.  Did you re-push?

Yes, that was commit https://github.com/CGAL/cgal/pull/3797/commits/7ffbf131bb615e19699680daa2e739ffb123d066.

**Comment by maxGimeno** [comment]
There are conflicts with master 
and I  strongly suspect this PR to be the origin of [those](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-34.shtml#Arrangement_on_surface_2).

**Comment by afabri** [comment]
> 
> 
> There are conflicts with master
> and I strongly suspect this PR to be the origin of [those](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-34.shtml#Arrangement_on_surface_2).

@lrineau took over the control of this branch.

**Comment by maxGimeno** [comment]
Still think the PR is responsible for the errors in the Arrangement testsuite on macOS

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-36.shtml.

**Comment by lrineau** [comment]
[**@maxGimeno**](https://github.com/maxGimeno) commented on [May 14, 2019, 4:03 PM GMT+2](https://github.com/CGAL/cgal/pull/3797#issuecomment-492251046 "2019-05-14T14:03:10Z - Replied by Github Reply Comments"):
> Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-36.shtml.

No, the result in Number_types are pretty bad:
https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-36.shtml#Number_types

**Comment by maxGimeno** [comment]
Yeah well I guess those columns were not in the master I used to compare. Or they were red over there too.

**Comment by maxGimeno** [comment]
Now it's all green in Number_types, so I ;m not sure, but it seems ok in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-39.shtml

File: Number_types/include/CGAL/boost_mp.h [hunk 4]
```diff
@@ -29,7 +29,7 @@
 // easy solution.
 // TODO: MSVC has trouble with versions <= 1.69, reenable once 1.70 has been
 // tested. https://github.com/boostorg/multiprecision/issues/98
-#if !defined CGAL_DO_NOT_USE_BOOST_MP && BOOST_VERSION >= 106300 && !defined _MSC_VER
+#if !defined CGAL_DISABLE_GMP && !defined CGAL_DO_NOT_USE_BOOST_MP && BOOST_VERSION >= 106300 && !defined _MSC_VER
```
Comment (mglisse) [inline Number_types/include/CGAL/boost_mp.h:32]: @lrineau Could you explain why?

Comment (afabri) [reply Number_types/include/CGAL/boost_mp.h:32]: My guess is that we only had the GMP-backend in mind for `boost::mp`.

Comment (lrineau) [reply Number_types/include/CGAL/boost_mp.h:32]: > @lrineau Could you explain why?

From what I can see, at that time I only considered `boost::multiprecision::mpz_int`, and not `cpp_int`, or whatever other backend.

Maybe there was a platform of the testsuite with compilation errors with `cpp_int`. Now that CGAL-5.2-beta1 is almost done, I think we can experiment again in `integration`: we are at the very beginning of the release cycle for CGAL-5.3 (due in June 2021). I suggest we remove the `!defined CGAL_DISABLE_GMP &&` and see what happens in the test results.

**Pull request merged**